sdks/python/apache_beam/runners/interactive/recording_manager.py

```diff
@@ -289,10 +291,16 @@ def _watch(self, pcolls):
     """
 
     watched_pcollections = set()
+    watched_dataframes = set()
     for watching in ie.current_env().watching():
       for _, val in watching:
         if isinstance(val, beam.pvalue.PCollection):
           watched_pcollections.add(val)
+        elif isinstance(val, DeferredBase):
+          watched_dataframes.add(val)
+    # Convert them all in a single step for efficiency.
+    for pcoll in to_pcollection(*watched_dataframes, always_return_tuple=True):
+      watched_pcollections.add(pcoll)
```

Comment:
I think it would be preferable to use `to_pcollection(yield='pandas')` here and then just concat the raw DataFrames we collect. That would avoid issues with the imperfect mapping back to beam schemas (e.g. pandas allows duplicate column names, and to_pcollection doesn't have a good story for mapping the index)

Reply:
Fur sure. I actually started down this road, but it got messier (still totally doable, but out of scope for my timebox) to deal with the count parameter and windowing info.

Reply:
Got it thanks - I filed BEAM-12246 for this.